File: frontend/src/sync/retryPolicy.ts
const BASE_DELAY_MS = 1000;
const MAX_DELAY_MS = 16000;
const MAX_RETRIES = 5;

export async function withRetry<T>(
  fn: () => Promise<T>,
  retries: number = MAX_RETRIES,
): Promise<T> {
  let attempt = 0;
  while (true) {
    try {
      return await fn();
    } catch (error) {
      attempt++;
      if (attempt >= retries) throw error;
      const delay = Math.min(BASE_DELAY_MS * Math.pow(2, attempt - 1), MAX_DELAY_MS);
      const jitter = delay * (0.5 + Math.random() * 0.5);
      await new Promise<void>(r => setTimeout(r, jitter));
    }
  }
}

File: ios/lib/sync/retry_policy.dart
import 'dart:math';

Future<T> withRetry<T>(
  Future<T> Function() fn, {
  int maxRetries = 5,
  Duration baseDelay = const Duration(milliseconds: 1000),
}) async {
  int attempt = 0;
  final random = Random();

  while (true) {
    try {
      return await fn();
    } catch (e) {
      attempt++;
      if (attempt >= maxRetries) rethrow;

      final exponentialDelay = baseDelay * pow(2, attempt - 1).toInt();
      final jitter = Duration(
        milliseconds: random.nextInt(exponentialDelay.inMilliseconds ~/ 2),
      );
      await Future.delayed(exponentialDelay + jitter);
    }
  }
}
